src/docs/memory-management.md

```diff
@@ -0,0 +1,103 @@
+
+// VWT->Destroy
+define linkonce_odr hidden ptr @"outlined destroy of output.VType"(ptr %0) #8 !dbg !143 {
+entry:
+  %.refType = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !144
+  %toDestroy = load ptr, ptr %.refType, align 8, !dbg !144
+  call void @swift_release(ptr %toDestroy) #1, !dbg !144
+  ret ptr %0, !dbg !144
+}
+```
+
+### Initializers
+
+When a reference type or a value type containing a reference property is passed as a parameter to a constructor, its reference count is retained before the call and released after:
+
+```swift
+var refType = RefType()
+var vtype = VType(refType: refType)
+```
+```
+%3 = call ptr @swift_retain(ptr returned %2) #4, !dbg !164
+store ptr %2, ptr %refType, align 8, !dbg !164
+call swiftcc void @"output.VType.init(refType: output.RefType) -> output.VType"(ptr noalias nocapture sret(%T6output5VTypeV) %vtype, ptr %2), !dbg !168
+%toDestroy = load ptr, ptr %refType, align 8, !dbg !170
+call void @swift_release(ptr %toDestroy) #1, !dbg !170
+```
+
+### Parameters
+
+When a reference type or a value type containing a reference property is passed as a parameter, its reference count is retained before the call and released after:
+
+```
+%1 = load ptr, ptr %vtype.refType, align 8, !dbg !154
+%2 = call ptr @swift_retain(ptr returned %1) #4, !dbg !154
+%.refType = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !156
+store ptr %1, ptr %.refType, align 8, !dbg !156
+call swiftcc void @"output.PassThrough(vtype: output.VType) -> ()"(ptr noalias nocapture dereferenceable(8) %0), !dbg !157
+%3 = call ptr @"outlined destroy of output.VType"(ptr %0), !dbg !158
+```
+
+### Return types
+
+When a value type is returned from a function, a new instance is created, and its reference count is retained before the return. The caller is responsible for releasing the reference:
+```
+%2 = load ptr, ptr %.refType, align 8, !dbg !173
+store ptr %2, ptr %vtype.debug, align 8, !dbg !175
+%3 = call ptr @swift_retain(ptr returned %2) #4, !dbg !176
+%.refType1 = getelementptr inbounds %T6output5VTypeV, ptr %0, i32 0, i32 0, !dbg !176
+store ptr %2, ptr %.refType1, align 8, !dbg !176
+```
+
+## Memory handling
+
+To handle native memory in the scenarios above, the projections should could use the value witness table to invoke `InitWithCopy` for copy operations and `Destroy` for finalization. These functions manage reference counts at any level of nesting.
+
+To ensure correct memory handling:
+ - Value types that contain reference properties should be projected as C# classes with destructors that call `Destroy`
+ - When types are marshalled into Swift, `InitWithCopy` should be invoked to create the copy/ `Destroy` should be called on the copy at the end of the block
+ - When types are marshalled from Swift, `Destroy` should be called at the end of the block
+ - When using the "copy" constructor on the C# side, `InitWithCopy` should be invoked
```

Comment:
Could you please exactly specify what is meant by "copy" constructor?
C# typically doesn't have such a term - structs can't really have it (it's implicit effectively and can't be controlled and or intercepted), classes typically don't want it (it's rare you need to create an exact copy of an existing class).

Is this referring to the `.ctor(IntPre swiftPayload)` ?

Reply:
I think we don't consider structs here as anything that has a ref type (hence need `InitWithCopy`) will be projected as class

Reply:
This is a private "copy" constructor on the C# side used for marshalling from Swift. I will clarify it.

Reply:
Updated, please take a look.